core/src/banking_trace.rs

```diff
@@ -0,0 +1,497 @@
+    #[error("IO Error: {0}")]
+    IoError(#[from] std::io::Error),
+
+    #[error("Serialization Error: {0}")]
+    SerializeError(#[from] bincode::Error),
+
+    #[error("Integer Cast Error: {0}")]
+    IntegerCastError(#[from] std::num::TryFromIntError),
+
+    #[error("Trace size is too small (must be larger than {1}): {0}")]
+    TooSmallTraceSize(u64, u64),
+}
+
+const BASENAME: &str = "events";
+const TRACE_FILE_ROTATE_COUNT: u64 = 14; // target 2 weeks retention under normal load
+const TRACE_FILE_WRITE_INTERVAL_MS: u64 = 100;
+const BUF_WRITER_CAPACITY: usize = 10 * 1024 * 1024;
+pub const TRACE_FILE_DEFAULT_ROTATE_BYTE_THRESHOLD: u64 = 1024 * 1024 * 1024;
+pub const EMPTY_BANKING_TRACE_SIZE: u64 = 0;
+pub const DEFAULT_BANKING_TRACE_SIZE: u64 =
+    TRACE_FILE_DEFAULT_ROTATE_BYTE_THRESHOLD * TRACE_FILE_ROTATE_COUNT;
+
+#[allow(clippy::type_complexity)]
+#[derive(Debug)]
+pub struct BankingTracer {
+    enabled_tracer: Option<(
+        Sender<TimedTracedEvent>,
+        Option<JoinHandle<TracerThreadResult>>,
+        Arc<AtomicBool>,
+    )>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct TimedTracedEvent(std::time::SystemTime, TracedEvent);
+
+#[derive(Serialize, Deserialize, Debug)]
+enum TracedEvent {
+    Bank(Slot, u32, BankStatus, usize),
+    PacketBatch(ChannelLabel, BankingPacketBatch),
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+enum BankStatus {
+    Started,
+    Ended,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, Copy)]
+pub enum ChannelLabel {
+    NonVote,
+    TpuVote,
+    GossipVote,
+    Dummy,
+}
+
+struct RollingConditionGrouped {
+    basic: RollingConditionBasic,
+    is_checked: bool,
+}
+
+impl RollingConditionGrouped {
+    fn new(basic: RollingConditionBasic) -> Self {
+        Self {
+            basic,
+            is_checked: bool::default(),
+        }
+    }
+
+    fn reset(&mut self) {
+        self.is_checked = false;
+    }
+}
+
+struct GroupedWriter<'a> {
+    now: DateTime<Local>,
+    underlying: &'a mut RollingFileAppender<RollingConditionGrouped>,
+}
+
+impl<'a> GroupedWriter<'a> {
+    fn new(underlying: &'a mut RollingFileAppender<RollingConditionGrouped>) -> Self {
+        Self {
+            now: Local::now(),
+            underlying,
+        }
+    }
+}
+
+impl RollingCondition for RollingConditionGrouped {
+    fn should_rollover(&mut self, now: &DateTime<Local>, current_filesize: u64) -> bool {
+        if !self.is_checked {
+            self.is_checked = true;
+            self.basic.should_rollover(now, current_filesize)
+        } else {
+            false
+        }
+    }
+}
+
+impl<'a> Write for GroupedWriter<'a> {
+    fn write(&mut self, buf: &[u8]) -> std::result::Result<usize, io::Error> {
+        self.underlying.write_with_datetime(buf, &self.now)
+    }
+    fn flush(&mut self) -> std::result::Result<(), io::Error> {
+        self.underlying.flush()
+    }
+}
+
+pub fn sender_overhead_minimized_receiver_loop<T, E, const SLEEP_MS: u64>(
```

Comment:
> Woah, did I miss const generics becoming stable!?

yep, it's been for awhile. :) the stabilization is minimal, though.

> name is confusing (to me), there isn't a sender in this fn directly (maybe in the on_recv), but that should be abstracted away from this naming imo.

yep. the naming was a bit hard because i wanted to include the nuance into the name to clearly indicate the justification of the existence of this helper `fn`.. how about this?:

```patch
diff --git a/core/src/banking_trace.rs b/core/src/banking_trace.rs
index ba441a01af..7fa8c9a35b 100644
--- a/core/src/banking_trace.rs
+++ b/core/src/banking_trace.rs
@@ -138,7 +138,7 @@ impl<'a> Write for GroupedWriter<'a> {
     }
 }

-pub fn sender_overhead_minimized_receiver_loop<T, E, const SLEEP_MS: u64>(
+pub fn receiving_loop_with_minimized_sender_overhead<T, E, const SLEEP_MS: u64>(
     exit: Arc<AtomicBool>,
     receiver: Receiver<T>,
     mut on_recv: impl FnMut(T) -> Result<(), E>,

```

Reply:
I think that name is much clearer 👍

Reply:
renamed :) e9674a1b3787163f1e2db8aa9a4b5c479c6df4a3